flink-runtime/src/main/scala/org/apache/flink/runtime/jobmanager/JobManager.scala

```diff
@@ -886,10 +957,42 @@ class JobManager(
       if (instanceManager.isRegistered(taskManager)) {
         log.info(s"Task manager ${taskManager.path} wants to disconnect, because $msg.")
 
-        instanceManager.unregisterTaskManager(taskManager, false)
+          instanceManager.unregisterTaskManager(taskManager, false)
         context.unwatch(taskManager)
       }
 
+    case msg: StopCluster =>
+
+      log.info(s"Stopping JobManager with final application status ${msg.finalStatus()} " +
+        s"and diagnostics: ${msg.message()}")
+
+      val respondTo = sender()
+
+      // stop all task managers
+      instanceManager.getAllRegisteredInstances.asScala foreach {
+        instance =>
+          instance.getActorGateway.tell(msg)
+      }
+
+      // send resource manager the ok
+      currentResourceManager match {
+        case Some(rm) =>
+
+          // inform rm
+          rm ! decorateMessage(msg)
+
+          respondTo ! decorateMessage(StopClusterSuccessful.get())
+
+          // trigger shutdown
+          shutdown()
+
+        case None =>
+          // retry
+          context.system.scheduler.scheduleOnce(
+            2 seconds, self, decorateMessage(msg)
+          )(context.dispatcher)
```

Comment:
You're right. There should be an upper bound for the number of retries. If there was never a RM, then is behaves the same.